File: src/services/whatsappService.ts
import axios from 'axios';

const EVOLUTION_API_URL = process.env.EVOLUTION_API_URL!;
const EVOLUTION_API_KEY = process.env.EVOLUTION_API_KEY!;

const isMock = !EVOLUTION_API_KEY || 
               EVOLUTION_API_KEY === 'sua_chave_secreta_aqui' || 
               process.env.MOCK_MODE === 'true';

const evolutionClient = isMock ? null : axios.create({
  baseURL: EVOLUTION_API_URL,
  headers: {
    apikey: EVOLUTION_API_KEY,
    'Content-Type': 'application/json',
  },
  timeout: 15000,
});

// Mock connection state that transitions from close -> connecting -> open
let mockConnectionState = 'close';

// ─── Tipos ────────────────────────────────────────────────────────────────────

export interface SendTextOptions {
  instanceName: string;
  to: string;
  text: string;
  delay?: number;
}

export interface CreateInstanceResult {
  instance: { instanceName: string; status: string };
  hash: { apikey: string };
  webhook?: { url: string; enabled: boolean };
  events?: string[];
}

// ─── Funções de Mensagem ──────────────────────────────────────────────────────

/**
 * Envia uma mensagem de texto para um número no WhatsApp.
 */
export async function sendTextMessage({
  instanceName,
  to,
  text,
  delay = 1200,
}: SendTextOptions): Promise<void> {
  if (isMock) {
    console.log(`[Mock WhatsApp] Sending message via '${instanceName}' to ${to} (delay: ${delay}ms):`);
    console.log(`> ${text}`);
    return;
  }
  await evolutionClient!.post(`/message/sendText/${instanceName}`, {
    number: to,
    text,
    delay,
    linkPreview: false,
  });
}

/**
 * Envia uma reação (emoji) a uma mensagem.
 */
export async function sendReaction(
  instanceName: string,
  to: string,
  messageId: string,
  emoji: string
): Promise<void> {
  if (isMock) {
    console.log(`[Mock WhatsApp] Sending reaction '${emoji}' to message '${messageId}' via '${instanceName}' to ${to}`);
    return;
  }
  await evolutionClient!.post(`/message/sendReaction/${instanceName}`, {
    key: { remoteJid: `${to}@s.whatsapp.net`, id: messageId },
    reaction: emoji,
  });
}

// ─── Gerenciamento de Instâncias ──────────────────────────────────────────────

/**
 * Cria uma nova instância do WhatsApp para um cliente.
 */
export async function createInstance(instanceName: string, webhookUrl: string): Promise<CreateInstanceResult> {
  if (isMock) {
    console.log(`[Mock WhatsApp] Creating instance '${instanceName}' with webhook '${webhookUrl}'`);
    mockConnectionState = 'close';
    return {
      instance: { instanceName, status: 'created' },
      hash: { apikey: 'mock_instance_key' }
    };
  }
  const { data } = await evolutionClient!.post('/instance/create', {
    instanceName,
    token: instanceName,
    qrcode: true,
    integration: 'WHATSAPP-BAILEYS',
    webhook: {
      url: webhookUrl,
      byEvents: true,
      base64: false,
      events: [
        'MESSAGES_UPSERT',
        'MESSAGES_UPDATE',
        'CONNECTION_UPDATE',
      ],
    },
  });
  return data;
}

/**
 * Obtém o QR Code de uma instância para o cliente escanear.
 */
export async function getQRCode(instanceName: string): Promise<{ base64: string; code: string } | null> {
  if (isMock) {
    console.log(`[Mock WhatsApp] Generating QR Code for instance '${instanceName}'`);
    // Simulated small green box / fake QR code base64
    const fakeQrBase64 = 'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAQAAAAEAAQMAAABFGoRRAAAABlBMVEUAAAD///+l2Z/dAAAAMklEQVR42mP8z8AARjDw/2fADn7+h0Ew+PkfBsHg538YBIOf/2EQDH7+h0Ew+PkfBsHgRwAAXJ/w98x27CgAAAAASUVORK5CYII=';
    mockConnectionState = 'connecting';
    return {
      base64: fakeQrBase64,
      code: 'mock_qr_code_string'
    };
  }
  try {
    const { data } = await evolutionClient!.get(`/instance/connect/${instanceName}`);
    return data;
  } catch {
    return null;
  }
}

/**
 * Verifica o status de conexão de uma instância.
 */
export async function getInstanceStatus(instanceName: string): Promise<'open' | 'close' | 'connecting' | null> {
  if (isMock) {
    // If it was close, transition to connecting. If it was connecting, transition to open.
    if (mockConnectionState === 'close') {
      mockConnectionState = 'connecting';
    } else if (mockConnectionState === 'connecting') {
      mockConnectionState = 'open';
    }
    return mockConnectionState as any;
  }
  try {
    const { data } = await evolutionClient!.get(`/instance/connectionState/${instanceName}`);
    return data?.instance?.state ?? null;
  } catch {
    return null;
  }
}

/**
 * Deleta uma instância.
 */
export async function deleteInstance(instanceName: string): Promise<void> {
  if (isMock) {
    console.log(`[Mock WhatsApp] Deleting instance '${instanceName}'`);
    mockConnectionState = 'close';
    return;
  }
  await evolutionClient!.delete(`/instance/delete/${instanceName}`);
}

File: AceleraBot.Api/Services/WhatsappService.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace AceleraBot.Api.Services;

public class QrCode
{
    public string Base64 { get; set; } = "";
    public string Code { get; set; } = "";
}

public interface IWhatsappService
{
    Task SendTextMessageAsync(string instanceName, string to, string text, int delay = 1200);
    Task CreateInstanceAsync(string instanceName, string webhookUrl);
    Task<QrCode?> GetQrCodeAsync(string instanceName);
    Task<string?> GetInstanceStatusAsync(string instanceName);
    Task DeleteInstanceAsync(string instanceName);
}

// Espelha src/services/whatsappService.ts (Evolution API).
public class WhatsappService : IWhatsappService
{
    private readonly IHttpClientFactory _factory;
    private readonly ILogger<WhatsappService> _log;
    private readonly bool _isMock;

    public WhatsappService(IHttpClientFactory factory, IConfiguration cfg, ILogger<WhatsappService> log)
    {
        _factory = factory;
        _log = log;
        var key = cfg["EVOLUTION_API_KEY"];
        _isMock = string.IsNullOrEmpty(key) || key == "sua_chave_secreta_aqui" || cfg["MOCK_MODE"] == "true";
    }

    private HttpClient Client() => _factory.CreateClient("evolution");

    public async Task SendTextMessageAsync(string instanceName, string to, string text, int delay = 1200)
    {
        if (_isMock) { _log.LogInformation("[Mock WhatsApp] -> {To}: {Text}", to, text); return; }
        var body = new { number = to, text, delay, linkPreview = false };
        await Client().PostAsJsonAsync($"/message/sendText/{instanceName}", body);
    }

    public async Task CreateInstanceAsync(string instanceName, string webhookUrl)
    {
        if (_isMock) { _log.LogInformation("[Mock WhatsApp] create instance {Inst}", instanceName); return; }
        var body = new
        {
            instanceName,
            token = instanceName,
            qrcode = true,
            integration = "WHATSAPP-BAILEYS",
            webhook = new
            {
                url = webhookUrl,
                byEvents = true,
                base64 = false,
                events = new[] { "MESSAGES_UPSERT", "MESSAGES_UPDATE", "CONNECTION_UPDATE" }
            }
        };
        await Client().PostAsJsonAsync("/instance/create", body);
    }

    public async Task<QrCode?> GetQrCodeAsync(string instanceName)
    {
        if (_isMock)
            return new QrCode { Base64 = "data:image/png;base64,iVBORw0KGgo=", Code = "mock_qr_code" };
        try
        {
            var res = await Client().GetAsync($"/instance/connect/{instanceName}");
            if (!res.IsSuccessStatusCode) return null;
            var json = await res.Content.ReadFromJsonAsync<JsonElement>();
            var b64 = json.TryGetProperty("base64", out var b) ? b.GetString() : null;
            var code = json.TryGetProperty("code", out var c) ? c.GetString() : null;
            if (b64 is null && code is null) return null;
            return new QrCode { Base64 = b64 ?? "", Code = code ?? "" };
        }
        catch { return null; }
    }

    public async Task<string?> GetInstanceStatusAsync(string instanceName)
    {
        if (_isMock) return "connecting";
        try
        {
            var res = await Client().GetAsync($"/instance/connectionState/{instanceName}");
            if (!res.IsSuccessStatusCode) return null;
            var json = await res.Content.ReadFromJsonAsync<JsonElement>();
            if (json.TryGetProperty("instance", out var inst) && inst.TryGetProperty("state", out var st))
                return st.GetString();
            return null;
        }
        catch { return null; }
    }

    public async Task DeleteInstanceAsync(string instanceName)
    {
        if (_isMock) { _log.LogInformation("[Mock WhatsApp] delete instance {Inst}", instanceName); return; }
        try { await Client().DeleteAsync($"/instance/delete/{instanceName}"); }
        catch (Exception e) { _log.LogError(e, "deleteInstance failed"); }
    }
}
